algorithms/linfa-tsne/src/lib.rs

```diff
@@ -0,0 +1,267 @@
+//! t-distributed stochastic neighbor embedding
+//!
+use ndarray::Array2;
+use ndarray_rand::rand::{rngs::SmallRng, Rng, SeedableRng};
+use ndarray_rand::rand_distr::Normal;
+
+use linfa::{dataset::DatasetBase, traits::Transformer, Float};
+
+mod error;
+pub use error::{Result, TSneError};
+
+/// The t-SNE algorithm is a statistical method for visualizing high-dimensional data by
+/// giving each datapoint a location in a two or three-dimensional map.
+///
+/// The t-SNE algorithm comprises two main stages. First, t-SNE constructs a probability
+/// distribution over pairs of high-dimensional objects in such a way that similar objects
+/// are assigned a higher probability while dissimilar points are assigned a lower probability.
+/// Second, t-SNE defines a similar probability distribution over the points in the low-dimensional
+/// map, and it minimizes the Kullback–Leibler divergence (KL divergence) between the two
+/// distributions with respect to the locations of the points in the map.
+///
+/// This crate wraps the [bhtsne](https://github.com/frjnn/bhtsne) crate for the linfa project. It
+/// implements the exact t-SNE, as well as the Barnes-Hut approximation.
+///
+/// # Examples
+///
+/// ```ignore
+/// use linfa::traits::Transformer;
+/// use linfa_tsne::TSne;
+///
+/// let ds = linfa_datasets::iris();
+///
+/// let ds = TSne::embedding_size(2)
+///     .perplexity(10.0)
+///     .approx_threshold(0.6)
+///     .transform(ds);
+/// ```
+pub struct TSne<F: Float, R: Rng + Clone> {
+    embedding_size: usize,
+    threshold: F,
+    perplexity: F,
+    max_iter: usize,
+    preliminary_iter: Option<usize>,
+    rng: R,
+}
+
+impl<F: Float> TSne<F, SmallRng> {
+    /// Create a t-SNE param set with given embedding size
+    ///
+    /// # Defaults to:
+    ///  * `theta`: 0.5
+    ///  * `perplexity`: 1.0
+    ///  * `max_iter`: 2000
+    ///  * `rng`: SmallRng with seed 42
+    pub fn embedding_size(embedding_size: usize) -> TSne<F, SmallRng> {
+        Self::embedding_size_with_rng(embedding_size, SmallRng::seed_from_u64(42))
+    }
+}
+
+impl<F: Float, R: Rng + Clone> TSne<F, R> {
+    /// Create a t-SNE param set with given embedding size and random number generator
+    ///
+    /// # Defaults to:
+    ///  * `theta`: 0.5
+    ///  * `perplexity`: 1.0
+    ///  * `max_iter`: 2000
+    pub fn embedding_size_with_rng(embedding_size: usize, rng: R) -> TSne<F, R> {
+        TSne {
+            embedding_size,
+            rng,
+            threshold: F::from(0.5).unwrap(),
+            perplexity: F::from(5.0).unwrap(),
```

Comment:
Perplexity is said to be 1.0 in the comments but is 5.0 in the code. Also applies to the comment for `embedding_size`

Reply:
Also, I don't know if it would be a good thing to specify that `theta` here refers to the `threshold` param (which is the name the user needs to use to change its value), mainly because I'm not really familiar with the algorithm so I don't know if they are used interchangeably.

Reply:
the parameter is called theta in the paper, but I have called it `approx_threshold` here because it's the ratio of cell size to distance at which the approximation is used.